src/sentry/snuba/discover.py

```diff
@@ -945,3 +947,141 @@ def get_facets(query, params, limit=10, referrer=None):
             )
 
     return results
+
+
+HistogramParams = namedtuple("HistogramParams", ["bucket_size", "start_offset", "multiplier"])
+
+
+def measurements_histogram_query(
+    measurements, user_query, params, num_buckets, precision=0, referrer=None,
+):
+    min_value, max_value = find_measurements_min_max(measurements, user_query, params)
+
+    key_col = "array_join(measurements_key)"
+    histogram_params = find_measurements_histogram_params(
+        num_buckets, min_value, max_value, precision
+    )
+    histogram_col = get_measurements_histogram_col(histogram_params)
+
+    results = query(
+        selected_columns=[key_col, histogram_col, "count()"],
+        conditions=[[get_function_alias(key_col), "IN", measurements]],
+        query=user_query,
+        params=params,
+        # the zerofill step assumes it is ordered by the bin then name
+        orderby=[get_function_alias(histogram_col), get_function_alias(key_col)],
+        limit=len(measurements) * num_buckets,
+        referrer=referrer,
+        auto_fields=True,
+        use_aggregate_conditions=True,
+    )
+
+    return normalize_measurements_histogram(
+        measurements, num_buckets, key_col, histogram_params, results
+    )
+
+
+def get_measurements_histogram_col(params):
+    return u"measurements_histogram({:d}, {:d}, {:d})".format(*params)
+
+
+def find_measurements_histogram_params(num_buckets, min_value, max_value, precision):
+    multiplier = int(10 ** precision)
+
+    # finding the bounds might result in None if there isn't sufficient data
+    if min_value is None or max_value is None:
+        return HistogramParams(1, 0, multiplier)
+
+    # A single bucket in a histogram contains values from [start, end),
+    # meaning that start is inclusive and end is exclusive
+
+    scaled_min = multiplier * min_value
+    scaled_max = multiplier * max_value
+
+    # align the first bin with the minimum value
+    start_offset = int(floor(scaled_min))
+
+    bucket_size = int(ceil((int(ceil(scaled_max)) - int(floor(scaled_min))) / float(num_buckets)))
```

Comment:
For my understanding why can't we just wrap this with one int instead of needing to do `int(ceil(` and `int(floor(`?

Reply:
Both the max and the min are floating point values (even after being scaled), so we don't want to truncate the range by casting to int at the end. So the `int(ceil(` and `int(floor(` is there to to make the max a little bigger and the min a little smaller.

Though it might just be me being paranoid and it might be fine without them. 💭

Reply:
My first impression is that this should be equivalent to using one int, but I might be missing something

```
int(ceil(int(ceil(5.xx) - int(floor(3.yy))))
= int(ceil(int(6.0) - int(3.0)))
= int(ceil(6-3))
= int(3.0)
```

vs
```
int(ceil(5.xx) - floor(3.yy))
=int(6.0 - 3.0)
=int(3.0)
```

Reply:
Addressed in 82d1acc